boa/src/builtins/string/mod.rs

```diff
@@ -479,39 +482,88 @@ impl String {
-                        let group_str = mat_caps.get(1).unwrap().as_str();
-                        let group_int = group_str.parse::<usize>().unwrap();
-                        result = re
-                            .replace(result.as_str(), caps.get(group_int).unwrap().as_str())
-                            .to_string()
+                    let mut result = StdString::new();
+                    let mut chars = val.chars().peekable();
+
+                    let m = caps.len();
+
+                    while let Some(first) = chars.next() {
+                        if first == '$' {
+                            let second = chars.next();
+                            let second_is_digit = second.map_or(false, |ch| ch.is_digit(10));
+                            // we use peek so that it is still in the iterator if not used
+                            let third = if second_is_digit { chars.peek() } else { None };
+                            let third_is_digit = third.map_or(false, |ch| ch.is_digit(10));
+
+                            match (second, third) {
+                                (Some('$'), _) => {
+                                    // $$
+                                    result.push('$');
+                                }
+                                (Some('&'), _) => {
+                                    // $&
+                                    let matched = caps.get(0).expect("cannot get matched value");
+                                    result.push_str(matched.as_str());
+                                }
+                                (Some('`'), _) => {
+                                    // $`
+                                    let start_of_match = mat.start();
+                                    result.push_str(&primitive_val[..start_of_match]);
+                                }
+                                (Some('\''), _) => {
+                                    // $'
+                                    let end_of_match = mat.end();
+                                    result.push_str(&primitive_val[end_of_match..]);
+                                }
+                                (Some(second), Some(third))
+                                    if second_is_digit && third_is_digit =>
+                                {
+                                    // $nn
+                                    let tens = second.to_digit(10).unwrap() as usize;
+                                    let units = third.to_digit(10).unwrap() as usize;
+                                    let nn = 10 * tens + units;
+                                    if nn == 0 || nn > m {
+                                        [Some(first), Some(second), chars.next()]
+                                            .iter()
+                                            .flatten()
+                                            .for_each(|ch: &char| result.push(*ch));
```

Comment:
```suggestion
                                        result.push(first);
                                        result.push(second);
                                        if Some(ch) = chars.next() {
                                           result.push(ch))
                                        }
```
Doing it like this should be more readable and more  efficient